indico/web/client/js/react/components/WTFDateTimeField.jsx

```diff
@@ -0,0 +1,238 @@
+// This file is part of Indico.
+// Copyright (C) 2002 - 2020 CERN
+//
+// Indico is free software; you can redistribute it and/or
+// modify it under the terms of the MIT License; see the
+// LICENSE file for more details.
+
+import 'react-dates/initialize';
+import TimePicker from 'rc-time-picker';
+import PropTypes from 'prop-types';
+import React, {useEffect, useMemo, useRef, useState, useCallback} from 'react';
+import moment from 'moment';
+import {SingleDatePicker} from 'indico/react/components';
+import {Translate} from 'indico/react/i18n';
+
+import 'rc-time-picker/assets/index.css';
+import 'react-dates/lib/css/_datepicker.css';
+
+function trigger(id) {
+  document.getElementById(id).dispatchEvent(new Event('change', {bubbles: true}));
+}
+
+export default function WTFDateTimeField({
+  timeId,
+  dateId,
+  timezoneFieldId,
+  timezone,
+  uses24HourFormat,
+  required,
+  disabled,
+  allowClear,
+  earliest,
+  latest,
+  defaultTime,
+  linkedField,
+}) {
+  const dateField = useMemo(() => document.getElementById(dateId), [dateId]);
+  const timeField = useMemo(() => document.getElementById(timeId), [timeId]);
+  const timezoneField = useMemo(() => document.getElementById(timezoneFieldId), [timezoneFieldId]);
+  const linkedFieldDateElem = useMemo(
+    () => linkedField && document.getElementById(`${linkedField.id}-datestorage`),
+    [linkedField]
+  );
+  const [time, setTime] = useState(timeField.value ? moment(timeField.value, 'HH:mm') : null);
```

Comment:
values are strings, so I think it'd be `''` instead of `null` with that change

Reply:
All right. I was expecting it to be undefined.

Reply:
I was reviewing this again and I'm not really a fan of the ternary operator here.
Moment will wrap the dates into an invalid state, so what do you think about using:
`const [time, setTime] = useState(moment(timeField.value, 'HH:mm'));`

And for checking if the time is valid we would use `time.isValid()` instead of `if time`.

Reply:
How would you get an invalid time though? It's coming from a hidden field populated by data coming from the server - so I think you can rely on it being either empty or valid.

And I'm pretty sure passing an invalid time to the timepicker will resulting it literally displaying "invalid time" or something like that, so you'd need another isValid check there...

Personally I'd be more likely to add something like this to clearly break in case of invalid data so it's not just resulting in possibly broken behavior;

    if (time && !time.isValid() { throw new Error('Invalid time'); )

Not sure if that's actually needed though... I guess in case of garbage date you'd still end up with a server-side validation error.

Reply:
Actually.. we already have this util:

```js
export function toMoment(dt, format) {
  return dt ? moment(dt, format) : null;
}
```

We could add a `strict` mode that throws in case of an invalid value and then use it like this:

```js
export function toMoment(dt, format, strict = false) {
  if (!dt) { return null; }
  const obj = moment(dt, format);
  if (!obj.isValid()) { throw new Error(`Invalid dt: ${dt}`); }
  return obj;
}
```

```js
const [time, setTime] = useState(toMoment(timeField.value, undefined, true));
```

Reply:
I agree with your suggestion too.
What I meant previously by invalid time was `''` or `null`. Either ways, `moment(timeField.value, 'HH:mm')` will be `.isValid() = false`.

In the event of using my previous suggestion, that would imply probably changing TimePicker to `value={time.isValid() || null}`.

Reply:
Yes, I guess we can't escape using it here.